ql/src/java/org/apache/hadoop/hive/ql/exec/HiveWritableComparator.java

```diff
@@ -0,0 +1,152 @@
+
+    @Override
+    public int compare(Object key1, Object key2) {
+        ArrayList a1 = (ArrayList) key1;
+        ArrayList a2 = (ArrayList) key2;
+        if (a1.size() != a2.size()) {
+            return a1.size() > a2.size() ? 1 : -1;
+        }
+        if (a1.size() == 0) {
+            return 0;
+        }
+
+        if (comparator == null) {
+            // For List, all elements should be of same type.
+            comparator = HiveWritableComparator.get(a1.get(0));
+        }
+
+        int result = 0;
+        for (int i = 0; i < a1.size(); i++) {
+            result = comparator.compare(a1.get(i), a2.get(i));
+            if (result != 0) {
+                return result;
+            }
+        }
+        return result;
+    }
+}
+
+class HiveStructComparator extends HiveWritableComparator {
+    HiveWritableComparator[] comparator = null;
+
+    @Override
+    public int compare(Object key1, Object key2) {
+        ArrayList a1 = (ArrayList) key1;
+        ArrayList a2 = (ArrayList) key2;
+        if (a1.size() != a2.size()) {
+            return a1.size() > a2.size() ? 1 : -1;
+        }
+        if (a1.size() == 0) {
+            return 0;
+        }
+        if (comparator == null) {
+            comparator = new HiveWritableComparator[a1.size()];
+            // For struct all elements may not be of same type, so create comparator for each entry.
+            for (int i = 0; i < a1.size(); i++) {
+                comparator[i] = HiveWritableComparator.get(a1.get(i));
+            }
+        }
+        int result = 0;
+        for (int i = 0; i < a1.size(); i++) {
+            result = comparator[i].compare(a1.get(i), a2.get(i));
+            if (result != 0) {
+                return result;
+            }
+        }
+        return result;
+    }
+}
+
+class HiveMapComparator extends HiveWritableComparator {
+    HiveWritableComparator comparatorValue = null;
+    HiveWritableComparator comparatorKey = null;
+
+    @Override
+    public int compare(Object key1, Object key2) {
+        LinkedHashMap map1 = (LinkedHashMap) key1;
+        LinkedHashMap map2 = (LinkedHashMap) key2;
+        if (map1.entrySet().size() != map2.entrySet().size()) {
+            return map1.entrySet().size() > map2.entrySet().size() ? 1 : -1;
+        }
+        if (map1.entrySet().size() == 0) {
+            return 0;
+        }
+
+        if (comparatorKey == null) {
+            comparatorKey = HiveWritableComparator.get(map1.keySet().iterator().next());
+            comparatorValue = HiveWritableComparator.get(map1.values().iterator().next());
+        }
+
+        int result = comparatorKey.compare(map1.keySet().iterator().next(),
+                map2.keySet().iterator().next());
+        if (result != 0) {
+            return result;
+        }
+        return comparatorValue.compare(map1.values().iterator().next(), map2.values().iterator().next());
+    }
+}
+
+public class HiveWritableComparator extends WritableComparator {
```

Comment:
Why is it necessary to introduce a new API? As far as I can see `HiveWritableComparator` does not add any new behavior to `WritableComparator`. It only contains some factory methods and these would fit much better in a `ComplexWritableComparatorFactory` class that is final and immutable. 

The non-public top-level classes above could become private static members classes of the factory class.

Reply:
done